<a href="https://colab.research.google.com/github/aasif8873/fake-news-detection/blob/main/fake.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
import re
import string

In [ ]:
data_fake = pd.read_csv('Fake.csv')
data_true = pd.read_csv('True.csv')

In [ ]:
data_fake['class'] =0
data_true['class'] = 1

In [ ]:
data_fake.head(10)

title  \
0   Donald Trump Sends Out Embarrassing New Year’...   
1   Drunk Bragging Trump Staffer Started Russian ...   
2   Sheriff David Clarke Becomes An Internet Joke...   
3   Trump Is So Obsessed He Even Has Obama’s Name...   
4   Pope Francis Just Called Out Donald Trump Dur...   
5   Racist Alabama Cops Brutalize Black Boy While...   
6   Fresh Off The Golf Course, Trump Lashes Out A...   
7   Trump Said Some INSANELY Racist Stuff Inside ...   
8   Former CIA Director Slams Trump Over UN Bully...   
9   WATCH: Brand-New Pro-Trump Ad Features So Muc...   

                                                text subject  \
0  Donald Trump just couldn t wish all Americans ...    News   
1  House Intelligence Committee Chairman Devin Nu...    News   
2  On Friday, it was revealed that former Milwauk...    News   
3  On Christmas day, Donald Trump announced that ...    News   
4  Pope Francis used his annual Christmas Day mes...    News   
5  The number of cases of cops brutalizing and ki...    News   
6  Donald Trump spent a good portion of his day a...    News   
7  In the wake of yet another court decision that...    News   
8  Many people have raised the alarm regarding th...    News   
9  Just when you might have thought we d get a br...    News   

                date  class  
0  December 31, 2017      0  
1  December 31, 2017      0  
2  December 30, 2017      0  
3  December 29, 2017      0  
4  December 25, 2017      0  
5  December 25, 2017      0  
6  December 23, 2017      0  
7  December 23, 2017      0  
8  December 22, 2017      0  
9  December 21, 2017      0

In [ ]:
data_true.head(10)

title  \
0  As U.S. budget fight looms, Republicans flip t...   
1  U.S. military to accept transgender recruits o...   
2  Senior U.S. Republican senator: 'Let Mr. Muell...   
3  FBI Russia probe helped by Australian diplomat...   
4  Trump wants Postal Service to charge 'much mor...   
5  White House, Congress prepare for talks on spe...   
6  Trump says Russia probe will be fair, but time...   
7  Factbox: Trump on Twitter (Dec 29) - Approval ...   
8         Trump on Twitter (Dec 28) - Global Warming   
9  Alabama official to certify Senator-elect Jone...   

                                                text       subject  \
0  WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1  WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2  WASHINGTON (Reuters) - The special counsel inv...  politicsNews   
3  WASHINGTON (Reuters) - Trump campaign adviser ...  politicsNews   
4  SEATTLE/WASHINGTON (Reuters) - President Donal...  politicsNews   
5  WEST PALM BEACH, Fla./WASHINGTON (Reuters) - T...  politicsNews   
6  WEST PALM BEACH, Fla (Reuters) - President Don...  politicsNews   
7  The following statements were posted to the ve...  politicsNews   
8  The following statements were posted to the ve...  politicsNews   
9  WASHINGTON (Reuters) - Alabama Secretary of St...  politicsNews   

                 date  class  
0  December 31, 2017       1  
1  December 29, 2017       1  
2  December 31, 2017       1  
3  December 30, 2017       1  
4  December 29, 2017       1  
5  December 29, 2017       1  
6  December 29, 2017       1  
7  December 29, 2017       1  
8  December 29, 2017       1  
9  December 28, 2017       1

In [ ]:
data_fake.shape, data_true.shape

((23481, 5), (21417, 5))

In [ ]:
data_fake_manual_testing = data_fake.tail(10)
max_index_fake = data_fake.index.max()
for i in range(max_index_fake, max_index_fake - 10, -1):
    data_fake.drop(i, axis=0, inplace=True)


data_true_manual_testing = data_true.tail(10)
max_index_true = data_true.index.max()
for i in range(max_index_true, max_index_true - 10, -1):
    data_true.drop(i, axis=0, inplace=True)

**Double-click (or enter) to edit**

In [ ]:
data_fake.shape, data_true.shape

((23461, 5), (21397, 5))

In [ ]:
data_fake_manual_testing['class'] = 0
data_true_manual_testing['class'] = 1

<ipython-input-15-90008d39c97b>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_fake_manual_testing['class'] = 0
<ipython-input-15-90008d39c97b>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_true_manual_testing['class'] = 1


In [ ]:
data_fake_manual_testing.head(10)

title  \
23461  REPORT: ‘Federal Government Escalated the Viol...   
23462  BOILER ROOM – Oregon Standoff, Cuddle Parties,...   
23463  Eyewitness Says Feds Ambushed Bundys, 100 Shot...   
23464  Episode #119 – SUNDAY WIRE: ‘You Know the Dril...   
23465  ‘There’ll be boots on the ground’: US making n...   
23466  Boston Brakes? How to Hack a New Car With Your...   
23467  Oregon Governor Says Feds ‘Must Act’ Against P...   
23468  Ron Paul on Burns Oregon Standoff and Jury Nul...   
23469     BOILER ROOM: As the Frogs Slowly Boil – EP #40   
23470  Arizona Rancher Protesting in Oregon is Target...   

                                                    text      subject  \
23461   KILLED: Rancher and protest spokesman Robert ...  Middle-east   
23462  Tune in to the Alternate Current Radio Network...  Middle-east   
23463  Patrick Henningsen 21st Century Wire UPDATE: 1...  Middle-east   
23464  Episode #119 of SUNDAY WIRE SHOW finally resum...  Middle-east   
23465  21st Century Wire says Various parties in Wash...  Middle-east   
23466  21st Century Wire says For those who still ref...  Middle-east   
23467  21st Century Wire says So far, after nearly 20...  Middle-east   
23468  21st Century Wire says If you ve been followin...  Middle-east   
23469  Tune in to the Alternate Current Radio Network...  Middle-east   
23470  RTOne of the most visible members of the armed...  Middle-east   

                   date  class  
23461  January 28, 2016      0  
23462  January 28, 2016      0  
23463  January 27, 2016      0  
23464  January 24, 2016      0  
23465  January 23, 2016      0  
23466  January 22, 2016      0  
23467  January 21, 2016      0  
23468  January 21, 2016      0  
23469  January 20, 2016      0  
23470  January 20, 2016      0

In [ ]:
data_true_manual_testing.head(10)

title  \
21397  Germany's Schulz says he would demand U.S. wit...   
21398  Blunt instrument? What a list of banned articl...   
21399  Saudi police release teenager detained for dan...   
21400  The People's Princess, Britons work to keep me...   
21401  Argentina labor unions protest job losses, Mac...   
21402  Exclusive: Trump's Afghan decision may increas...   
21403  U.S. puts more pressure on Pakistan to help wi...   
21404  Exclusive: U.S. to withhold up to $290 million...   
21405  Trump talks tough on Pakistan's 'terrorist' ha...   
21406  U.S., North Korea clash at U.N. forum over nuc...   

                                                    text    subject  \
21397  BERLIN (Reuters) - The leader of Germany s Soc...  worldnews   
21398  SHANGHAI (Reuters) - An old review of an acade...  worldnews   
21399  DUBAI (Reuters) - A 14-year-old boy who was de...  worldnews   
21400  LONDON (Reuters) - Abdul Daoud spilt most of t...  worldnews   
21401  BUENOS AIRES (Reuters) - Argentina s main labo...  worldnews   
21402  ON BOARD A U.S. MILITARY AIRCRAFT (Reuters) - ...  worldnews   
21403  WASHINGTON (Reuters) - The United States sugge...  worldnews   
21404  WASHINGTON (Reuters) - The United States has d...  worldnews   
21405  ISLAMABAD (Reuters) - Outlining a new strategy...  worldnews   
21406  GENEVA (Reuters) - North Korea and the United ...  worldnews   

                   date  class  
21397  August 23, 2017       1  
21398  August 23, 2017       1  
21399  August 22, 2017       1  
21400  August 23, 2017       1  
21401  August 22, 2017       1  
21402  August 22, 2017       1  
21403  August 21, 2017       1  
21404  August 22, 2017       1  
21405  August 22, 2017       1  
21406  August 22, 2017       1

In [ ]:
data_merge =pd.concat([data_fake, data_true], axis =0)
data_merge.head(10)

title  \
0   Donald Trump Sends Out Embarrassing New Year’...   
1   Drunk Bragging Trump Staffer Started Russian ...   
2   Sheriff David Clarke Becomes An Internet Joke...   
3   Trump Is So Obsessed He Even Has Obama’s Name...   
4   Pope Francis Just Called Out Donald Trump Dur...   
5   Racist Alabama Cops Brutalize Black Boy While...   
6   Fresh Off The Golf Course, Trump Lashes Out A...   
7   Trump Said Some INSANELY Racist Stuff Inside ...   
8   Former CIA Director Slams Trump Over UN Bully...   
9   WATCH: Brand-New Pro-Trump Ad Features So Muc...   

                                                text subject  \
0  Donald Trump just couldn t wish all Americans ...    News   
1  House Intelligence Committee Chairman Devin Nu...    News   
2  On Friday, it was revealed that former Milwauk...    News   
3  On Christmas day, Donald Trump announced that ...    News   
4  Pope Francis used his annual Christmas Day mes...    News   
5  The number of cases of cops brutalizing and ki...    News   
6  Donald Trump spent a good portion of his day a...    News   
7  In the wake of yet another court decision that...    News   
8  Many people have raised the alarm regarding th...    News   
9  Just when you might have thought we d get a br...    News   

                date  class  
0  December 31, 2017      0  
1  December 31, 2017      0  
2  December 30, 2017      0  
3  December 29, 2017      0  
4  December 25, 2017      0  
5  December 25, 2017      0  
6  December 23, 2017      0  
7  December 23, 2017      0  
8  December 22, 2017      0  
9  December 21, 2017      0

In [ ]:
data_merge.columns

Index(['title', 'text', 'subject', 'date', 'class'], dtype='object')

In [ ]:
data = data_merge.drop(['title', 'subject', 'date'], axis =1)

In [ ]:
data.isnull().sum()

text     0
class    0
dtype: int64

In [ ]:
data = data.sample(frac = 1)

In [ ]:
data.head()

text  class
17191  BEIJING (Reuters) - The  One China  principle ...      1
22998  21st Century Wire says The Oroville dam in Nor...      0
19888  As our limp-wristed President flies around on ...      0
3015   A member of the Idaho House of Representatives...      0
14336  Next up is a law suit to make it legal for H-1...      0

In [ ]:
data.reset_index(inplace = True)
data.drop(['index'], axis =1, inplace = True)

In [ ]:
data.columns

Index(['text', 'class'], dtype='object')

In [ ]:
data.head()

text  class
0  BEIJING (Reuters) - The  One China  principle ...      1
1  21st Century Wire says The Oroville dam in Nor...      0
2  As our limp-wristed President flies around on ...      0
3  A member of the Idaho House of Representatives...      0
4  Next up is a law suit to make it legal for H-1...      0

In [ ]:
def wordopt(text):
    if isinstance(text, str):
        text = text.lower()
        text = re.sub('\[.*?\]', '', text)
        text = re.sub("\\W", " ", text)
        text = re.sub('https?://\S+|www\.\S+', '', text)
        text = re.sub('<.*?>+', '', text)
        text = re.sub('[%s]' % re.escape(string.punctuation), '', text)
        text = re.sub('\n', '', text)
        text = re.sub('\w*\d\w*', '', text)
    else:
        text = ''
    return text

In [ ]:
data['text'] = data['text'].apply(wordopt)

In [ ]:
x = data['text']
y = data['class']

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.25)

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorization = TfidfVectorizer()
xv_train = vectorization.fit_transform(x_train)
xv_test = vectorization.transform(x_test)

In [ ]:
from sklearn.linear_model import LogisticRegression

LR = LogisticRegression()
LR.fit(xv_train, y_train)

LogisticRegression()

In [ ]:
pred_lr = LR.predict(xv_test)

In [ ]:
LR.score(xv_test, y_test)

0.9866250557289344

In [ ]:
print(classification_report(y_test, pred_lr))

              precision    recall  f1-score   support

           0       0.99      0.99      0.99      5904
           1       0.98      0.99      0.99      5311

    accuracy                           0.99     11215
   macro avg       0.99      0.99      0.99     11215
weighted avg       0.99      0.99      0.99     11215



In [ ]:
from sklearn.tree import DecisionTreeClassifier

DT = DecisionTreeClassifier()
DT.fit(xv_train, y_train)

DecisionTreeClassifier()

In [ ]:
pred_dt = DT.predict(xv_test)

In [ ]:
DT.score(xv_test, y_test)

0.9950958537672759

In [ ]:
print(classification_report(y_test, pred_dt))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00      5904
           1       0.99      1.00      0.99      5311

    accuracy                           1.00     11215
   macro avg       1.00      1.00      1.00     11215
weighted avg       1.00      1.00      1.00     11215



In [ ]:
from sklearn.ensemble import GradientBoostingClassifier

GB = GradientBoostingClassifier(random_state = 0)
GB.fit(xv_train, y_train)

GradientBoostingClassifier(random_state=0)

In [ ]:
pred_gb = GB.predict(xv_test)

In [ ]:
GB.score(xv_test, y_test)

0.9954525189478377

In [ ]:
print(classification_report(y_test, pred_gb))

              precision    recall  f1-score   support

           0       1.00      0.99      1.00      5904
           1       0.99      1.00      1.00      5311

    accuracy                           1.00     11215
   macro avg       1.00      1.00      1.00     11215
weighted avg       1.00      1.00      1.00     11215



In [ ]:
from sklearn.ensemble import RandomForestClassifier
RF = RandomForestClassifier(random_state = 0)
RF.fit(xv_train, y_train)

RandomForestClassifier(random_state=0)

In [ ]:
pred_rf = RF.predict(xv_test)

In [ ]:
RF.score(xv_test, y_test)

0.9894783771734285

In [ ]:
print(classification_report(y_test, pred_rf))

              precision    recall  f1-score   support

           0       0.99      0.99      0.99      5904
           1       0.99      0.99      0.99      5311

    accuracy                           0.99     11215
   macro avg       0.99      0.99      0.99     11215
weighted avg       0.99      0.99      0.99     11215



In [ ]:
def output_lable(n):
    if n == 0:
        return "Fake News"
    elif n == 1:
        return "Not A Fake News"

def manual_testing(news):
   testing_news = {"text":[news]}
   new_def_test = pd.DataFrame(testing_news)
   new_def_test["text"] = new_def_test["text"].apply(wordopt)
   new_x_test = new_def_test["text"]
   new_xv_test = vectorization.transform(new_x_test)
   pred_LR = LR.predict(new_xv_test)
   pred_DT = DT.predict(new_xv_test)
   pred_GB = GB.predict(new_xv_test)
   pred_RF = RF.predict(new_xv_test)

   return print("\n\nLR Prediction: {} \nDT Prediction: {} \nGB Prediction: {} \nRF Prediction: {}".format(output_lable(pred_LR[0]),
                                                                                                            output_lable(pred_DT[0]),
                                                                                                          output_lable(pred_GB[0]),
                                                                                                          output_lable(pred_RF[0])))

In [ ]:
news = str(input())
manual_testing(news)

Union Minister Shivraj Chouhan hits out at Air India for ‘broken’ seat; airline apologises


LR Prediction: Fake News 
DT Prediction: Fake News 
GB Prediction: Fake News 
RF Prediction: Fake News


In [ ]:
news = str(input())
manual_testing(news)

Boeing will likely miss a key 737 MAX jet production target in 2024, analysts at rating agencies Moody's and S&P told Reuters, saying that the company faces challenges as it ramps up its strongest-selling plane.  The US aerospace giant's goal is to produce 38 MAX jets a month by the end of 2024, up from 25 jets a month in July. Click here to connect with us on WhatsApp But Moody's and S&P said that goal may not be reached until 2025 due to risks like possible labor disruptions at the planemaker's facilities in the Seattle area. Boeing, however, faces no immediate risk of a credit downgrade that would drop its rating to junk levels, the two rating agencies said on Friday.  Jonathan Root, lead Boeing analyst at Moody's, assumes the planemaker will end 2024 producing 32 MAX jets per month, and reach the target of 38 in the second half of 2025. "We remain in a 'show me' state of mind," he said.  MAX production and deliveries, which are closely watched by investors and airlines, mostly slow

**test with this input for better outcome**

Boeing will likely miss a key 737 MAX jet production target in 2024, analysts at rating agencies Moody's and S&P told Reuters, saying that the company faces challenges as it ramps up its strongest-selling plane.

The US aerospace giant's goal is to produce 38 MAX jets a month by the end of 2024, up from 25 jets a month in July. Click here to connect with us on WhatsApp But Moody's and S&P said that goal may not be reached until 2025 due to risks like possible labor disruptions at the planemaker's facilities in the Seattle area. Boeing, however, faces no immediate risk of a credit downgrade that would drop its rating to junk levels, the two rating agencies said on Friday.

Jonathan Root, lead Boeing analyst at Moody's, assumes the planemaker will end 2024 producing 32 MAX jets per month, and reach the target of 38 in the second half of 2025. "We remain in a 'show me' state of mind," he said.

MAX production and deliveries, which are closely watched by investors and airlines, mostly slowed following a Jan. 5 mid-air panel blowout on a new 737 MAX 9 that exposed longstanding quality-control problems at the jetmaker.

Boeing slowed output so it could improve production quality, but the decline in output and deliveries has taken a toll on cash flow. It burned about $8.3 billion in cash in the first half of 2024 and expects free cash flow to be negative this year, burdening its balance sheet.

"We see risks to reaching that number (38), including labor negotiations and the company’s history of underdelivering relative to targets," said Ben Tsocanos, aerospace director at S&P Global Ratings. "We view increasing and stabilizing MAX production as necessary to generating free cash flow, which is ultimately what we care about to maintain the rating." Both S&P and Moody's rate Boeing one notch above junk status.